src/Management/src/Endpoint/Actuators/HttpExchanges/HttpExchangesDiagnosticObserver.cs

```diff
@@ -0,0 +1,216 @@
+    {
+        ArgumentNullException.ThrowIfNull(optionsMonitor);
+
+        _optionsMonitor = optionsMonitor;
+        _logger = loggerFactory.CreateLogger<HttpExchangesDiagnosticObserver>();
+    }
+
+    public HttpExchangesResult GetHttpExchanges()
+    {
+        List<HttpExchange> recentExchanges = [];
+        HttpExchangesEndpointOptions options = _optionsMonitor.CurrentValue;
+
+        recentExchanges.AddRange(
+            from exchange in Queue
+            let requestUri = FilterUri(exchange.Request.Uri, options)
+            let filteredRequestHeaders = options.IncludeRequestHeaders ? FilterHeaders(exchange.Request.Headers, options.RequestHeaders) : []
+            let filteredResponseHeaders = options.IncludeResponseHeaders ? FilterHeaders(exchange.Response.Headers, options.ResponseHeaders) : []
+            let principal = options.IncludeUserPrincipal ? exchange.Principal : null
+            let remoteAddress = options.IncludeRemoteAddress ? exchange.Request.RemoteAddress : null
+            let session = options.IncludeSessionId ? exchange.Session : null
+            let request = new HttpExchangeRequest(exchange.Request.Method, requestUri, filteredRequestHeaders, remoteAddress)
+            let response = new HttpExchangeResponse(exchange.Response.Status, filteredResponseHeaders)
+            select new HttpExchange(request, response, exchange.Timestamp, principal, session, exchange.TimeTaken));
+
+        if (options.Reverse)
+        {
+            recentExchanges = recentExchanges.OrderByDescending(exchange => exchange.Timestamp).ToList();
+        }
+
+        return new HttpExchangesResult(recentExchanges);
+    }
+
+    public override void ProcessEvent(string eventName, object? value)
+    {
+        if (eventName != StopEventName || value == null)
+        {
+            return;
+        }
+
+        Activity? current = Activity.Current;
+
+        if (current == null)
+        {
+            return;
+        }
+
+        HttpContext? context = GetHttpContextPropertyValue(value);
+
+        if (context != null)
+        {
+            RecordHttpExchange(current, context);
+        }
+    }
+
+    internal static HttpContext? GetHttpContextPropertyValue(object instance)
+    {
+        return GetPropertyOrDefault<HttpContext>(instance, "HttpContext");
+    }
+
+    private void RecordHttpExchange(Activity current, HttpContext context)
+    {
+        ArgumentNullException.ThrowIfNull(current);
+        ArgumentNullException.ThrowIfNull(context);
+
+        HttpExchange trace = GetHttpExchange(context, current.Duration);
+        Queue.Enqueue(trace);
+
+        if (Queue.Count > _optionsMonitor.CurrentValue.Capacity && !Queue.TryDequeue(out _))
+        {
+            _logger.LogDebug("Stop - Dequeue failed");
+        }
+    }
+
+    private static Uri FilterUri(Uri requestUri, HttpExchangesEndpointOptions options)
+    {
+        var filteredUri = new UriBuilder(requestUri);
```

Comment:
I think we should redact user/password here. See `UriExtensions.ToMaskedUri`.

Reply:
adding test `HttpExchangesActuator_DoesNotCaptureAuthInUri`